docs/config.rst
Configuration
=============

Several options exist that allow you to configure the behavior of the
``Raven_Client``. These are passed as the second parameter of the
constructor, and is expected to be an array of key value pairs:

.. code-block:: php

    $client = new Raven_Client($dsn, array(
        'option_name' => 'value',
    ));

Available Settings
------------------

The following settings are available for the client:

.. describe:: name

    A string to override the default value for the server's hostname.

    Defaults to ``Raven_Compat::gethostname()``.

.. describe:: tags

    An array of tags to apply to events in this context.

    .. code-block:: php

        'tags' => array(
            'php_version' => phpversion(),
        )

    .. code-block:: php

        $client->tags_context(array(
            'php_version' => phpversion(),
        ));

.. describe:: release

    The version of your application (e.g. git SHA)

    .. code-block:: php

        'release' => MyApp::getReleaseVersion(),

    .. code-block:: php

        $client->setRelease(MyApp::getReleaseVersion());

.. describe:: environment

    The environment your application is running in.

    .. code-block:: php

        'environment' => 'production',

    .. code-block:: php

        $client->setEnvironment('production');

.. describe:: app_path

    The root path to your application code.

    .. code-block:: php

        'app_path' => app_root(),

    .. code-block:: php

        $client->setAppPath(app_root());

.. describe:: excluded_app_paths

    Paths to exclude from app_path detection.

    .. code-block:: php

        'excluded_app_paths' => array(app_root() . '/cache'),

    .. code-block:: php

        $client->setExcludedAppPaths(array(app_root() . '/cache'));

.. describe:: prefixes

    Prefixes which should be stripped from filenames to create relative
    paths.

    .. code-block:: php

        'prefixes' => array(
            '/www/php/lib',
        ),

    .. code-block:: php

        $client->setPrefixes(array(
            '/www/php/lib',
        ));

.. describe:: sample_rate

    The sampling factor to apply to events. A value of 0.00 will deny sending
    any events, and a value of 1.00 will send 100% of events.

    .. code-block:: php

        // send 50% of events
        'sample_rate' => 0.5,

.. describe:: send_callback

    A function which will be called whenever data is ready to be sent. Within
    the function you can mutate the data, or alternatively return ``false`` to
    instruct the SDK to not send the event.

    .. code-block:: php

        'send_callback' => function($data) {
            // strip HTTP data
            @unset($data['request']);
        },

    .. code-block:: php

        $client->setSendCallback(function($data) {
            // dont send events if POST
            if ($_SERVER['REQUEST_METHOD'] === 'POST')
            {
                return false;
            }
        });

.. describe:: curl_method

    Defaults to 'sync'.

    Available methods:

    - ``sync`` (default): send requests immediately when they're made
    - ``async``: uses a curl_multi handler for best-effort asynchronous
      submissions
    - ``exec``: asynchronously send events by forking a curl
      process for each item

.. describe:: curl_path

    Defaults to 'curl'.

    Specify the path to the curl binary to be used with the 'exec' curl
    method.

.. describe:: transport

    Set a custom transport to override how Sentry events are sent upstream.

    .. code-block:: php

        'transport' => function($client, $data) {
            $myHttpClient->send(array(
                'url'     => $client->getServerEndpoint(),
                'method'  => 'POST',
                'headers' => array(
                    'Content-Encoding' => 'gzip',
                    'Content-Type'     => 'application/octet-stream',
                    'User-Agent'       => $client->getUserAgent(),
                    'X-Sentry-Auth'    => $client->getAuthHeader(),
                ),
                'body'    => gzcompress(jsonEncode($data)),
            ))
        },

    .. code-block:: php

        $client->setTransport(...);

.. describe:: trace

    Set this to ``false`` to disable reflection tracing (function calling
    arguments) in stacktraces.


.. describe:: logger

    Adjust the default logger name for messages.

    Defaults to ``php``.

.. describe:: ca_cert

    The path to the CA certificate bundle.

    Defaults to the common bundle which includes getsentry.com:
    ./data/cacert.pem

    Caveats:

    - The CA bundle is ignored unless curl throws an error suggesting it
      needs a cert.
    - The option is only currently used within the synchronous curl
      transport.

.. describe:: curl_ssl_version

    The SSL version (2 or 3) to use.  By default PHP will try to determine
    this itself, although in some cases this must be set manually.

.. describe:: message_limit

    Defaults to 1024 characters.

    This value is used to truncate message and frame variables. However it
    is not guarantee that length of whole message will be restricted by
    this value.

.. describe:: processors

    An array of classes to use to process data before it is sent to
    Sentry. By default, ``Raven_Processor_SanitizeDataProcessor`` is used

.. describe:: processorOptions

    Options that will be passed on to a ``setProcessorOptions()`` function
    in a ``Raven_Processor`` sub-class before that Processor is added to
    the list of processors used by ``Raven_Client``

    An example of overriding the regular expressions in
    ``Raven_Processor_SanitizeDataProcessor`` is below:

    .. code-block:: php

        'processorOptions' => array(
            'Raven_Processor_SanitizeDataProcessor' => array(
                        'fields_re' => '/(user_password|user_token|user_secret)/i',
                        'values_re' => '/^(?:\d[ -]*?){15,16}$/'
                    )
        )

.. describe:: timeout

    The timeout for sending requests to the Sentry server in seconds, default is 2 seconds.

    .. code-block:: php

        'timeout' => 2,

.. describe:: excluded_exceptions

    Exception that should not be reported, exceptions extending exceptions in this list will also
    be excluded, default is an empty array.

    In the example below, when you exclude ``LogicException`` you will also exclude ``BadFunctionCallException``
    since it extends ``LogicException``.

    .. code-block:: php

        'excluded_exceptions' => array('LogicException'),

.. describe:: ignore_server_port

    By default the server port will be added to the logged URL when it is a non
    standard port (80, 443).
    Setting this to ``true`` will ignore the server port altogether and will
    result in the server port never getting appended to the logged URL.

    .. code-block:: php

        'ignore_server_port' => true,

.. _sentry-php-request-context:

Providing Request Context
-------------------------

Most of the time you're not actually calling out to Raven directly, but
you still want to provide some additional context. This lifecycle
generally constists of something like the following:

- Set some context via a middleware (e.g. the logged in user)
- Send all given context with any events during the request lifecycle
- Cleanup context

There are three primary methods for providing request context:

.. code-block:: php

    // bind the logged in user
    $client->user_context(array('email' => 'foo@example.com'));

    // tag the request with something interesting
    $client->tags_context(array('interesting' => 'yes'));

    // provide a bit of additional context
    $client->extra_context(array('happiness' => 'very'));


If you're performing additional requests during the lifecycle, you'll also
need to ensure you cleanup the context (to reset its state):

.. code-block:: php

    $client->context->clear();
    
Processors
----------

The following processors are available bundled with sentry-php. They can be used in ``processors`` configuration, and configured through ``processorOptions`` as described above.

.. describe:: Raven_Processor_SanitizeDataProcessor

    This is the default processor. It replaces fields or values with asterisks
    in frames, http, and basic extra data.
   
   Available options:
   
   - ``fields_re``: takes a regex expression of fields to sanitize
     Defaults to ``/(authorization|password|passwd|secret|password_confirmation|card_number|auth_pw)/i``
   - ``values_re``: takes a regex expression of values to sanitize
     Defaults to ``/^(?:\d[ -]*?){13,16}$/``

.. describe:: Raven_Processor_SanitizeHttpHeadersProcessor

   This processor sanitizes the configured HTTP headers to ensure no sensitive
   information is sent to the server.
   
   Available options:
   
   - ``sanitize_http_headers``: takes an array of headers to sanitize. 
     Defaults to ``['Authorization', 'Proxy-Authorization', 'X-Csrf-Token', 'X-CSRFToken', 'X-XSRF-TOKEN']``

.. describe:: Raven_Processor_SanitizeStacktraceProcessor

   This processor removes the `pre_context`, `context_line` and `post_context`
   information from all exceptions captured by an event.

.. describe:: Raven_Processor_RemoveHttpBodyProcessor

   This processor removes all the data of the HTTP body to ensure no sensitive
   information is sent to the server in case the request method is POST, PUT,
   PATCH or DELETE.
 
.. describe:: Raven_Processor_RemoveCookiesProcessor
 
   This processor removes all the cookies from the request to ensure no sensitive
   information is sent to the server.


CHANGES
1.7.0
-----

- Corrected some issues with argument serialization in stacktraces (#399).
- The default exception handler will now re-raise exceptions when
  ``call_existing`` is true and no exception handler is registered (#421).
- Collect User.ip_address automatically (#419).

1.6.2
-----

- Fixed behavior where fatal errors weren't correctly being reported in most situations.

1.6.1
-----

- Correct handling of null in ``user_context``.

1.6.0
-----

- Improved serialization of certain types to be more restrictive.
- ``error_types`` can now be configured via ``RavenClient``.
- Class serialization has been expanded to include attributes.
- The session extension is no longer required.
- Monolog is no longer a required dependency.
- ``user_context`` now merges by default.

1.5.0
-----

- Added named transaction support.

1.4.0
-----

This version primarily overhauls the exception/stacktrace generation to fix
a few bugs and improve the quality of data (#359).

- Added ``excluded_app_paths`` config.
- Removed ``shift_vars`` config.
- Correct fatal error handling to only operate on expected types. This also
  fixes some behavior with the error suppression operator.
- Expose anonymous and similar frames in the stacktrace.
- Default ``prefixes`` to PHP's include paths.
- Remove ``module`` usage.
- Better handle empty argument context.
- Correct alignment of filename (current frame) and function (caller frame)

1.3.0
-----

- Fixed an issue causing the error suppression operator to not be respected (#335)
- Fixed some serialization behavior (#352)
- Fixed an issue with app paths and trailing slashes (#350)
- Handle non-latin encoding with source code context line (#345)

1.2.0
-----

- Handle non-latin encoding in source code and exception values (#342)
- Ensure pending events are sent on shutdown by default (#338)
- Add ``captureLastError`` helper (#334)
- Dont report duplicate errors with fatal error handler (#334)
- Enforce maximum length for string serialization (#329)

1.1.0
-----

- Uncoercable values should no longer prevent exceptions from sending
  to the Sentry server.
- ``install()`` can no longer be called multiple times.

1.0.0
-----

- Removed deprecated error codes configuration from ErrorHandler.
- Removed env data from HTTP interface.
- Removed 'message' attribute from exceptions'.
- appPath and prefixes are now resolved fully.
- Fixed various getter methods requiring invalid args.
- Fixed data mutation with 'send_callback'.

0.22.0
------

- Improve handling of encodings.
- Improve resiliency of variable serialization.
- Add 'formatted' attribute to Message interface.

0.21.0
------

- Added ``transport`` option.
- Added ``install()`` shortcut.

0.20.0
------

- Handle missing function names on frames.
- Remove suppression operator usage in breadcrumbs buffer.
- Force serialization of context values.

0.19.0
------

- Add error_reporting breadcrumb handler.

0.18.0
------

- Remove session from serialized data.
- send_callback return value must now be false to prevent capture.
- Add various getter/setter methods for configuration.

0.17.0
------

- Don't attempt to serialize fixed SDK inputs.
- Improvements to breadcrumbs support in Monolog.

0.16.0
------

- Initial breadcrumbs support with Monolog handler.

0.15.0
------

- Fixed some cases where serialization wouldn't happen.
- Added sdk attribute.

0.14.0
------

- Added ``prefixes`` option for stripping absolute paths.
- Removed ``abs_path`` from stacktraces.
- Added ``app_path`` to specify application root for resolving ``in_app` on frames.
- Moved Laravel support to ``sentry-laravel`` project.
- Fixed duplicate stack computation.
- Added ``dsn`` option to ease configuration.
- Fixed an issue with the curl async transport.
- Improved serialization of values.

0.13.0
------

- Updated API to use new style interfaces.
- Remove session cookie in default processor.
- Expand docs for Laravel, Symfony2, and Monolog.
- Default error types can now be set as part of ErrorHandler configuration.

0.12.1
------

- Dont send empty values for various context.

0.12.0
------

- Bumped protocol version to 6.
- Fixed an issue with the async curl handler (GH-216).
- Removed UDP transport.

0.11.0
------

- New configuration parameter: 'release'
- New configuration parameter: 'message_limit'
- New configuration parameter: 'curl_ssl_version'
- New configuration parameter: 'curl_ipv4'
- New configuration parameter: 'verify_ssl'
- Updated remote endpoint to use modern project-based path.
- Expanded default sanitizer support to include 'auth_pw' attribute.

0.10.0
------

- Added a default certificate bundle which includes common root CA's
  as well as getsentry.com's CA.

0.9.1
-----

- Change default curl connection to 'sync'
- Improve CLI reporting


0.9.0
-----

- Protocol version 5
- Default to asynchronous HTTP handler using curl_multi.


(For previous versions see the commit history)


lib/Raven/Configuration.php
<?php

/*
 * This file is part of Raven.
 *
 * (c) Sentry Team
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Raven;

use Symfony\Component\OptionsResolver\Options;
use Symfony\Component\OptionsResolver\OptionsResolver;

/**
 * Configuration container for the Sentry client.
 *
 * @author Stefano Arlandini <sarlandini@alice.it>
 */
class Configuration
{
    /**
     * @var array The configuration options
     */
    private $options = [];

    /**
     * @var string|null A simple server string, set to the DSN found on your Sentry settings
     */
    private $dsn;

    /**
     * @var string The project ID number to send to the Sentry server
     */
    private $projectId;

    /**
     * @var string The public key to authenticate the SDK
     */
    private $publicKey;

    /**
     * @var string The secret key to authenticate the SDK
     */
    private $secretKey;

    /**
     * @var OptionsResolver The options resolver
     */
    private $resolver;

    /**
     * Class constructor.
     *
     * @param array $options The configuration options
     */
    public function __construct(array $options = [])
    {
        $this->resolver = new OptionsResolver();

        $this->configureOptions($this->resolver);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the number of attempts to resend an event that failed to be sent.
     *
     * @return int
     */
    public function getSendAttempts()
    {
        return $this->options['send_attempts'];
    }

    /**
     * Sets the number of attempts to resend an event that failed to be sent.
     *
     * @param int $attemptsCount The number of attempts
     */
    public function setSendAttempts($attemptsCount)
    {
        $options = array_merge($this->options, ['send_attempts' => $attemptsCount]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the prefixes which should be stripped from filenames to create
     * relative paths.
     *
     * @return string[]
     */
    public function getPrefixes()
    {
        return $this->options['prefixes'];
    }

    /**
     * Sets the prefixes which should be stripped from filenames to create
     * relative paths.
     *
     * @param array $prefixes The prefixes
     */
    public function setPrefixes(array $prefixes)
    {
        $options = array_merge($this->options, ['prefixes' => $prefixes]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets whether all the objects should be serialized.
     *
     * @return bool
     */
    public function getSerializeAllObjects()
    {
        return $this->options['serialize_all_object'];
    }

    /**
     * Sets whether all the objects should be serialized.
     *
     * @param bool $serializeAllObjects Flag indicating if all objects should be serialized
     */
    public function setSerializeAllObjects($serializeAllObjects)
    {
        $options = array_merge($this->options, ['serialize_all_object' => $serializeAllObjects]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the sampling factor to apply to events. A value of 0 will deny
     * sending any events, and a value of 1 will send 100% of events.
     *
     * @return float
     */
    public function getSampleRate()
    {
        return $this->options['sample_rate'];
    }

    /**
     * Sets the sampling factor to apply to events. A value of 0 will deny
     * sending any events, and a value of 1 will send 100% of events.
     *
     * @param float $sampleRate The sampling factor
     */
    public function setSampleRate($sampleRate)
    {
        $options = array_merge($this->options, ['sample_rate' => $sampleRate]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the character encoding detection order.
     *
     * @return string[]|null
     */
    public function getMbDetectOrder()
    {
        return $this->options['mb_detect_order'];
    }

    /**
     * Sets the character encoding detection order.
     *
     * @param string[]|null $detectOrder The detection order
     */
    public function setMbDetectOrder($detectOrder)
    {
        $options = array_merge($this->options, ['mb_detect_order' => $detectOrder]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets whether the stacktrace must be auto-filled.
     *
     * @return bool
     */
    public function getAutoLogStacks()
    {
        return $this->options['auto_log_stacks'];
    }

    /**
     * Sets whether the stacktrace must be auto-filled.
     *
     * @param bool $enable Flag indicating if the stacktrace must be auto-filled
     */
    public function setAutoLogStacks($enable)
    {
        $options = array_merge($this->options, ['auto_log_stacks' => $enable]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the number of lines of code context to capture, or null if none.
     *
     * @return int|null
     */
    public function getContextLines()
    {
        return $this->options['context_lines'];
    }

    /**
     * Sets the number of lines of code context to capture, or null if none.
     *
     * @param int|null $contextLines The number of lines of code
     */
    public function setContextLines($contextLines)
    {
        $options = array_merge($this->options, ['context_lines' => $contextLines]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the encoding type for event bodies (GZIP or JSON).
     *
     * @return string
     */
    public function getEncoding()
    {
        return $this->options['encoding'];
    }

    /**
     * Sets the encoding type for event bodies (GZIP or JSON).
     *
     * @param string $encoding The encoding type
     */
    public function setEncoding($encoding)
    {
        $options = array_merge($this->options, ['encoding' => $encoding]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the current environment.
     *
     * @return string
     */
    public function getCurrentEnvironment()
    {
        return $this->options['current_environment'];
    }

    /**
     * Sets the current environment.
     *
     * @param string $environment The environment
     */
    public function setCurrentEnvironment($environment)
    {
        $options = array_merge($this->options, ['current_environment' => $environment]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the whitelist of environments that will send notifications to
     * Sentry.
     *
     * @return string[]
     */
    public function getEnvironments()
    {
        return $this->options['environments'];
    }

    /**
     * Sets the whitelist of environments that will send notifications to
     * Sentry.
     *
     * @param string[] $environments The environments
     */
    public function setEnvironments(array $environments)
    {
        $options = array_merge($this->options, ['environments' => $environments]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the list of logger 'progname's to exclude from breadcrumbs.
     *
     * @return string[]
     */
    public function getExcludedLoggers()
    {
        return $this->options['excluded_loggers'];
    }

    /**
     * Sets the list of logger 'progname's to exclude from breadcrumbs.
     *
     * @param string[] $loggers The list of logger 'progname's
     */
    public function setExcludedLoggers(array $loggers)
    {
        $options = array_merge($this->options, ['excluded_loggers' => $loggers]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the list of exception classes that should be ignored when sending
     * events to Sentry.
     *
     * @return string[]
     */
    public function getExcludedExceptions()
    {
        return $this->options['excluded_exceptions'];
    }

    /**
     * Sets the list of exception classes that should be ignored when sending
     * events to Sentry.
     *
     * @param string[] $exceptions The list of exception classes
     */
    public function setExcludedExceptions(array $exceptions)
    {
        $options = array_merge($this->options, ['excluded_exceptions' => $exceptions]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Checks whether the given exception should be ignored when sending events
     * to Sentry.
     *
     * @param \Throwable|\Exception $exception The exception
     *
     * @return bool
     */
    public function isExcludedException($exception)
    {
        foreach ($this->options['excluded_exceptions'] as $exceptionClass) {
            if ($exception instanceof $exceptionClass) {
                return true;
            }
        }

        return false;
    }

    /**
     * Gets the list of paths to exclude from app_path detection.
     *
     * @return string[]
     */
    public function getExcludedProjectPaths()
    {
        return $this->options['excluded_app_paths'];
    }

    /**
     * Sets the list of paths to exclude from app_path detection.
     *
     * @param array $paths The list of paths
     */
    public function setExcludedProjectPaths(array $paths)
    {
        $options = array_merge($this->options, ['excluded_app_paths' => $paths]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the project ID number to send to the Sentry server.
     *
     * @return string
     */
    public function getProjectId()
    {
        return $this->projectId;
    }

    /**
     * Gets the project which the authenticated user is bound to.
     *
     * @return string|null
     */
    public function getProjectRoot()
    {
        return $this->options['project_root'];
    }

    /**
     * Sets the project which the authenticated user is bound to.
     *
     * @param string|null $path The path to the project root
     */
    public function setProjectRoot($path)
    {
        $options = array_merge($this->options, ['project_root' => $path]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the public key to authenticate the SDK.
     *
     * @return string|null
     */
    public function getPublicKey()
    {
        return $this->publicKey;
    }

    /**
     * Gets the secret key to authenticate the SDK.
     *
     * @return string|null
     */
    public function getSecretKey()
    {
        return $this->secretKey;
    }

    /**
     * Gets the logger used by Sentry.
     *
     * @return string
     */
    public function getLogger()
    {
        return $this->options['logger'];
    }

    /**
     * Sets the logger used by Sentry.
     *
     * @param string $logger The logger
     */
    public function setLogger($logger)
    {
        $options = array_merge($this->options, ['logger' => $logger]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the release tag to be passed with every event sent to Sentry.
     *
     * @return string
     */
    public function getRelease()
    {
        return $this->options['release'];
    }

    /**
     * Sets the release tag to be passed with every event sent to Sentry.
     *
     * @param string $release The release
     */
    public function setRelease($release)
    {
        $options = array_merge($this->options, ['release' => $release]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets the DSN of the Sentry server the authenticated user is bound to.
     *
     * @return string
     */
    public function getDsn()
    {
        return $this->dsn;
    }

    /**
     * Gets the name of the server the SDK is running on (e.g. the hostname).
     *
     * @return string
     */
    public function getServerName()
    {
        return $this->options['server_name'];
    }

    /**
     * Sets the name of the server the SDK is running on (e.g. the hostname).
     *
     * @param string $serverName The server name
     */
    public function setServerName($serverName)
    {
        $options = array_merge($this->options, ['server_name' => $serverName]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Checks whether all events or a specific event (if provided) are allowed
     * to be captured.
     *
     * @param Event|null $event The event object
     *
     * @return bool
     */
    public function shouldCapture(Event $event = null)
    {
        $result = true;

        if (!empty($this->options['environments']) && !in_array($this->options['current_environment'], $this->options['environments'])) {
            $result = false;
        }

        if (null !== $this->options['should_capture'] && null !== $event) {
            $result = $result && $this->options['should_capture']($event);
        }

        return $result;
    }

    /**
     * Sets an optional callable to be called to decide whether an event should
     * be captured or not.
     *
     * @param callable|null $callable The callable
     */
    public function setShouldCapture(callable $callable = null)
    {
        $options = array_merge($this->options, ['should_capture' => $callable]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets a list of default tags for events.
     *
     * @return string[]
     */
    public function getTags()
    {
        return $this->options['tags'];
    }

    /**
     * Sets a list of default tags for events.
     *
     * @param string[] $tags A list of tags
     */
    public function setTags(array $tags)
    {
        $options = array_merge($this->options, ['tags' => $tags]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Gets a bit mask for error_reporting used in {@link ErrorHandler::handleError}.
     *
     * @return int
     */
    public function getErrorTypes()
    {
        return $this->options['error_types'];
    }

    /**
     * Sets a bit mask for error_reporting used in {@link ErrorHandler::handleError}.
     *
     * @param int $errorTypes The bit mask
     */
    public function setErrorTypes($errorTypes)
    {
        $options = array_merge($this->options, ['error_types' => $errorTypes]);

        $this->options = $this->resolver->resolve($options);
    }

    /**
     * Configures the options of the client.
     *
     * @param OptionsResolver $resolver The resolver for the options
     *
     * @throws \Symfony\Component\OptionsResolver\Exception\UndefinedOptionsException
     * @throws \Symfony\Component\OptionsResolver\Exception\AccessException
     */
    private function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults([
            'send_attempts' => 6,
            'prefixes' => explode(PATH_SEPARATOR, get_include_path()),
            'serialize_all_object' => false,
            'sample_rate' => 1,
            'mb_detect_order' => null,
            'auto_log_stacks' => true,
            'context_lines' => 3,
            'encoding' => 'gzip',
            'current_environment' => 'default',
            'environments' => [],
            'excluded_loggers' => [],
            'excluded_exceptions' => [],
            'excluded_app_paths' => [],
            'project_root' => null,
            'logger' => 'php',
            'release' => null,
            'dsn' => isset($_SERVER['SENTRY_DSN']) ? $_SERVER['SENTRY_DSN'] : null,
            'server_name' => gethostname(),
            'should_capture' => null,
            'tags' => [],
            'error_types' => null,
        ]);

        $resolver->setAllowedTypes('send_attempts', 'int');
        $resolver->setAllowedTypes('prefixes', 'array');
        $resolver->setAllowedTypes('serialize_all_object', 'bool');
        $resolver->setAllowedTypes('sample_rate', ['int', 'float']);
        $resolver->setAllowedTypes('mb_detect_order', ['null', 'array']);
        $resolver->setAllowedTypes('auto_log_stacks', 'bool');
        $resolver->setAllowedTypes('context_lines', 'int');
        $resolver->setAllowedTypes('encoding', 'string');
        $resolver->setAllowedTypes('current_environment', 'string');
        $resolver->setAllowedTypes('environments', 'array');
        $resolver->setAllowedTypes('excluded_loggers', 'array');
        $resolver->setAllowedTypes('excluded_exceptions', 'array');
        $resolver->setAllowedTypes('excluded_app_paths', 'array');
        $resolver->setAllowedTypes('project_root', ['null', 'string']);
        $resolver->setAllowedTypes('logger', 'string');
        $resolver->setAllowedTypes('release', ['null', 'string']);
        $resolver->setAllowedTypes('dsn', ['null', 'boolean', 'string']);
        $resolver->setAllowedTypes('server_name', 'string');
        $resolver->setAllowedTypes('should_capture', ['null', 'callable']);
        $resolver->setAllowedTypes('tags', 'array');
        $resolver->setAllowedTypes('error_types', ['null', 'int']);

        $resolver->setAllowedValues('encoding', ['gzip', 'json']);
        $resolver->setAllowedValues('dsn', function ($value) {
            switch (strtolower($value)) {
                case '':
                case 'false':
                case '(false)':
                case 'empty':
                case '(empty)':
                case 'null':
                case '(null)':
                    return true;
            }

            $parsed = @parse_url($value);

            if (false === $parsed) {
                return false;
            }

            if (!isset($parsed['scheme'], $parsed['user'], $parsed['host'], $parsed['path'])) {
                return false;
            }

            if (empty($parsed['user']) || (isset($parsed['pass']) && empty($parsed['pass']))) {
                return false;
            }

            if (!in_array(strtolower($parsed['scheme']), ['http', 'https'])) {
                return false;
            }

            return true;
        });

        $resolver->setNormalizer('dsn', function (Options $options, $value) {
            switch (strtolower($value)) {
                case '':
                case 'false':
                case '(false)':
                case 'empty':
                case '(empty)':
                case 'null':
                case '(null)':
                    $this->dsn = null;

                    return null;
            }

            $parsed = @parse_url($value);

            $this->dsn = $parsed['scheme'] . '://' . $parsed['host'];

            if (isset($parsed['port']) && ((80 !== $parsed['port'] && 'http' === $parsed['scheme']) || (443 !== $parsed['port'] && 'https' === $parsed['scheme']))) {
                $this->dsn .= ':' . $parsed['port'];
            }

            $this->dsn .= substr($parsed['path'], 0, strripos($parsed['path'], '/'));
            $this->publicKey = $parsed['user'];
            $this->secretKey = isset($parsed['pass']) ? $parsed['pass'] : null;

            $parts = explode('/', $parsed['path']);

            $this->projectId = array_pop($parts);

            return $value;
        });

        $resolver->setNormalizer('project_root', function (Options $options, $value) {
            if (null === $value) {
                return null;
            }

            return $this->normalizeAbsolutePath($value);
        });

        $resolver->setNormalizer('prefixes', function (Options $options, $value) {
            return array_map([$this, 'normalizeAbsolutePath'], $value);
        });

        $resolver->setNormalizer('excluded_app_paths', function (Options $options, $value) {
            return array_map([$this, 'normalizeAbsolutePath'], $value);
        });
    }

    /**
     * Normalizes the given path as an absolute path.
     *
     * @param string $value The path
     *
     * @return string
     */
    private function normalizeAbsolutePath($value)
    {
        $path = @realpath($value);

        if (false === $path) {
            $path = $value;
        }

        if (
            \DIRECTORY_SEPARATOR === substr($path, 0, 1)
            && \DIRECTORY_SEPARATOR !== substr($path, -1)
            && '.php' !== substr($path, -4)
        ) {
            $path .= \DIRECTORY_SEPARATOR;
        }

        return $path;
    }
}
